### PR TITLE
[JAV-544] metrics core implement (only queue metrics collected)

### PR DESCRIPTION
**Pull request by zhengyangyong**
Signed-off-by: zhengyangyong <yangyong.zheng@huawei.com>

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://https://servicecomb.atlassian.net/projects/JAV/issues) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[JAV-XXX] Fixes bug in ApproximateQuantiles`, where you replace `JAV-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
Main task is JAV-544 Reflector Metrics Architecture
also include JAV-509 User can custom Java Chassis Metrics and JAV-539 Need improve TpsAngLatency & QueueMetrics Calculation (***TpsAngLatency will be later**)

### CODE REVIEW

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/EmbeddedMetricsName.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+
+package io.servicecomb.metrics.core;
+
+public class EmbeddedMetricsName {
+
+  private static final String QUEUE_HEADER_FORMAT = "servicecomb.%s.queue";
+
+  public static final String QUEUE_COUNT_IN_QUEUE = QUEUE_HEADER_FORMAT + ".waitInQueue.count";
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/EmbeddedMetricsName.java]: why 2 QUEUE in the name? maybe `COUNT_IN_QUEUE_TEMPLATE`?

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/EmbeddedMetricsName.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.servicecomb.metrics.core;
+
+public class EmbeddedMetricsName {
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/EmbeddedMetricsName.java]: preferable name: EmbeddedMetricTemplates

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationStartProcessingEventListener.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    String countInQueueName = String.format(EmbeddedMetricsName.QUEUE_COUNT_IN_QUEUE, event.getOperationName());
+    String lifeTimeInQueueName = String.format(EmbeddedMetricsName.QUEUE_LIFE_TIME_IN_QUEUE, event.getOperationName());
+
+    Metric metric = registry.getMetric(countInQueueName);
+    if (metric == null) {
+      metric = registry.getOrCreateMetric(factory.createCounter(countInQueueName));
+    }
+    String instanceCountInQueueName = String.format(EmbeddedMetricsName.QUEUE_COUNT_IN_QUEUE, "instance");
+    Metric instanceMetric = registry.getMetric(instanceCountInQueueName);
+    metric.update(-1);
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationStartProcessingEventListener.java]: why -1? preferably `metric.decrement()`

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationStartedEventListener.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    String countInQueueName = String.format(EmbeddedMetricsName.QUEUE_COUNT_IN_QUEUE, event.getOperationName());
+    Metric metric = registry.getMetric(countInQueueName);
+    if (metric == null) {
+      metric = registry.getOrCreateMetric(factory.createCounter(countInQueueName));
+    }
+    String instanceCountInQueueName = String.format(EmbeddedMetricsName.QUEUE_COUNT_IN_QUEUE, "instance");
+    Metric instanceMetric = registry.getMetric(instanceCountInQueueName);
+
+    metric.update(1);
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationStartedEventListener.java]: preferably `metric.increment()`

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/BasicTimerMetric.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+      extraTagMetrics.put(key, monitor);
+    }
+
+    Monitor average = new Monitor() {
+      @Override
+      public Object getValue() {
+        double count = get("count").doubleValue();
+        if (count != 0) {
+          return get("totalTime").doubleValue() / get("count").doubleValue();
+        } else {
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/BasicTimerMetric.java]: else clause is not necessary

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/BasicTimerMetric.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+      String key = monitor.getConfig().getTags().getValue("statistic");
+      extraTagMetrics.put(key, monitor);
+    }
+
+    Monitor average = new Monitor() {
+      @Override
+      public Object getValue() {
+        double count = get("count").doubleValue();
+        if (count != 0) {
+          return get("totalTime").doubleValue() / get("count").doubleValue();
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/BasicTimerMetric.java]: why not directly use `count` variable in division?

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/BasicTimerMetric.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        return (Number) extraTagMetrics.get(tag).getValue(0);
+      } else {
+        throw new ServiceCombException("can't find tag in " + getName() + " metric");
+      }
+    }
+  }
+
+  @Override
+  public Map<String, Number> getAll() {
+    Map<String, Number> values = new HashMap<>(6);
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/BasicTimerMetric.java]: why 6? hash map has a default load factor of 0.75. this will cause it to rehash

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/CustomMetric.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+  private final Func0<Number> getCallback;
+
+  public CustomMetric(String name, Func0<Number> getCallback) {
+    super(name);
+    this.getCallback = getCallback;
+  }
+
+  @Override
+  public void update(Number num) {
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/CustomMetric.java]: seems this is a readonly metric. is it possible to use different interfaces for readonly & read/write metrics?

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/EmbeddedMetricsName.java (hunk 1 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/EmbeddedMetricsName.java]: Fixed

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/EmbeddedMetricsName.java (hunk 2 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/EmbeddedMetricsName.java]: Fixed

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationStartProcessingEventListener.java (hunk 3 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationStartProcessingEventListener.java]: Fixed

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationStartedEventListener.java (hunk 4 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationStartedEventListener.java]: Fixed

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/BasicTimerMetric.java (hunk 6 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/BasicTimerMetric.java]: Fixed

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/BasicTimerMetric.java (hunk 5 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/BasicTimerMetric.java]: Fixed

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/BasicTimerMetric.java (hunk 7 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/BasicTimerMetric.java]: Fixed

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/CustomMetric.java (hunk 8 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/CustomMetric.java]: add a new writable metric interface extends from metric interface

File: core/src/main/java/io/servicecomb/core/metrics/InvocationFinishedEvent.java [hunk 9]
```diff
@@ -29,6 +29,18 @@
 
   private final String operationType;
```
Comment (wujimin) [inline core/src/main/java/io/servicecomb/core/metrics/InvocationFinishedEvent.java]: in core project, we can use InvocationType

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationFinishedEventListener.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    this.factory = factory;
+  }
+
+  @Override
+  public Class<? extends MetricsEvent> getConcernedEvent() {
+    return InvocationFinishedEvent.class;
+  }
+
+  @Override
+  public void process(MetricsEvent data) {
```
Comment (wujimin) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationFinishedEventListener.java]: during invocation, it's better only change statistics

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationFinishedEventListener.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    InvocationFinishedEvent event = (InvocationFinishedEvent) data;
+
+    String executionTimeName = String.format(EmbeddedMetricTemplates.EXECUTION_TIME_TEMPLATE, event.getOperationName());
```
Comment (wujimin) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationFinishedEventListener.java]: format is not necessary

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationFinishedEventListener.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    WritableMetric metric = (WritableMetric) registry.getMetric(executionTimeName);
+    if (metric == null) {
+      metric = (WritableMetric) registry.getOrCreateMetric(factory.createTimer(executionTimeName));
```
Comment (wujimin) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationFinishedEventListener.java]: concurrent problem

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationFinishedEventListener.java (hunk 11 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationFinishedEventListener.java]: Operation是动态的，需要自动创建，对应的Metrics名需要format一下

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationFinishedEventListener.java (hunk 10 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationFinishedEventListener.java]: 没有看懂

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationFinishedEventListener.java (hunk 12 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationFinishedEventListener.java]: Fixed

File: core/src/main/java/io/servicecomb/core/metrics/InvocationFinishedEvent.java (hunk 9 above)
Comment (zhengyangyong) [reply core/src/main/java/io/servicecomb/core/metrics/InvocationFinishedEvent.java]: Fixed

**Comment by zhengyangyong** [comment]
rebase on latest master

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/registry/ThreadLocalMonitorManager.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    if (monitor == null) {
+      monitor = new InvocationThreadLocalMonitor();
+      allThreadMonitor.add(monitor);
+      localMonitor.set(monitor);
+    }
+    return monitor;
+  }
+
+  public static List<InvocationThreadLocalCache> getAllInvocationThreadLocalCache() {
+    List<InvocationThreadLocalCache> models = new ArrayList<>();
```
Comment (zhengyangyong) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/registry/ThreadLocalMonitorManager.java]: i will change "models" to "caches"

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/registry/InvocationThreadLocalCache.java [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+    if (isValidMin(executionTimeMin, localCache.executionTimeMin)) {
+      executionTimeMin = localCache.executionTimeMin;
+    }
+    if (isValidMax(executionTimeMax, localCache.executionTimeMax)) {
+      executionTimeMax = localCache.executionTimeMax;
+    }
+  }
+
+  public InvocationThreadLocalCache collect() {
+    InvocationThreadLocalCache clone = new InvocationThreadLocalCache(operationName, countInQueue, providerCallCount,
```
Comment (zhengyangyong) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/registry/InvocationThreadLocalCache.java]: i will change "clone" to "cloneCache"

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/model/RegistryModel.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+package io.servicecomb.metrics.core.model;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class RegistryModel {
+
+  private InstanceModel instanceModel;
```
Comment (zhengyangyong) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/model/RegistryModel.java]: forget finnal

File: metrics/metrics-core/src/test/java/io/servicecomb/metrics/core/TestEventAndRunner.java [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+    EventUtils
+        .triggerEvent(new InvocationFinishedEvent("fun12", InvocationType.PRODUCER, System.nanoTime(),
+            TimeUnit.MILLISECONDS.toNanos(600), TimeUnit.MILLISECONDS.toNanos(1100)));
+
+    EventUtils.triggerEvent(new InvocationStartedEvent("fun11", System.nanoTime()));
+
+    DefaultStatisticsRunner runner = new DefaultStatisticsRunner();
+
+    //we need some time escape for sim statistics interval
+    Thread.sleep(100);
```
Comment (WillemJiang) [inline metrics/metrics-core/src/test/java/io/servicecomb/metrics/core/TestEventAndRunner.java]: we don't need to wait here.

File: metrics/metrics-core/src/test/java/io/servicecomb/metrics/core/TestEventAndRunner.java [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+    runner.run();
+
+    RegistryModel model = runner.getRegistryModel();
+
+    Assert.assertTrue(model.getInvocations().get("fun1").getCountInQueue() == 0);
```
Comment (WillemJiang) [inline metrics/metrics-core/src/test/java/io/servicecomb/metrics/core/TestEventAndRunner.java]: Please use assertEquals instead of assertTrue to help the user to find out the reason of the test failure.

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/model/InstanceModel.java [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.servicecomb.metrics.core.model;
+
+public class InstanceModel {
```
Comment (WillemJiang) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/model/InstanceModel.java]: I cannot get the what's this class doing by reading its name. It looks like about some static information about the Queue.

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/model/InvocationModel.java [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.servicecomb.metrics.core.model;
+
+public class InvocationModel {
```
Comment (WillemJiang) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/model/InvocationModel.java]: Please use the base class to hold common value to the static module, to avoid writing the same code again.

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/registry/ThreadLocalMonitorManager.java (hunk 13 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/registry/ThreadLocalMonitorManager.java]: Done

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/registry/InvocationThreadLocalCache.java (hunk 14 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/registry/InvocationThreadLocalCache.java]: Done

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/model/RegistryModel.java (hunk 15 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/model/RegistryModel.java]: Done

File: metrics/metrics-core/src/test/java/io/servicecomb/metrics/core/TestEventAndRunner.java (hunk 16 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/test/java/io/servicecomb/metrics/core/TestEventAndRunner.java]: Deleted

File: metrics/metrics-core/src/test/java/io/servicecomb/metrics/core/TestEventAndRunner.java (hunk 17 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/test/java/io/servicecomb/metrics/core/TestEventAndRunner.java]: Done

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/model/InstanceModel.java (hunk 18 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/model/InstanceModel.java]: Done

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/model/InvocationModel.java (hunk 19 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/model/InvocationModel.java]: Done

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationStartedEventListener.java [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+
+  @Override
+  public Class<? extends Event> getConcernedEvent() {
+    return InvocationStartedEvent.class;
+  }
+
+  @Override
+  public void process(Event data) {
+    InvocationStartedEvent event = (InvocationStartedEvent) data;
+    registry.getRegistryMonitor().getInvocationMonitor(event.getOperationName()).getWaitInQueue().increment();
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationStartedEventListener.java]: this event listen seems to know too much about the internal of registry...

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/TimeMetric.java [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.servicecomb.metrics.core.metric;
+
+public class TimeMetric {
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/TimeMetric.java]: the name is misleading. it doesn't seem to be related to time

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/TimeMetric.java [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+    this.max = max;
+  }
+
+  public TimeMetric merge(TimeMetric metric) {
+    return new TimeMetric(this.total + metric.total, this.count + metric.count,
+        getMin(this.min, metric.min), getMax(this.max, metric.max));
+  }
+
+  private long getMin(long value1, long value2) {
+    return value1 == 0 || (value2 != 0 && value2 < value1) ? value2 : value1;
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/TimeMetric.java]: why needing comparison with 0?

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/TimeMetric.java [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  private long getMax(long value1, long value2) {
+    return value1 == 0 || value2 > value1 ? value2 : value1;
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/TimeMetric.java]: why missing a comparison with 0 like getMin?

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/monitor/RegistryMonitor.java [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+public class RegistryMonitor {
+
+  private final Map<String, InvocationMonitor> invocationMonitors;
+
+  public RegistryMonitor() {
+    this.invocationMonitors = new ConcurrentHashMap<>();
+  }
+
+  public InvocationMonitor getInvocationMonitor(String operationName) {
+    InvocationMonitor level = invocationMonitors.getOrDefault(operationName, null);
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/monitor/RegistryMonitor.java]: this `getOrDefault` is unnecessary because of `computeIfAbsent` below

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/monitor/TimeMonitor.java [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+package io.servicecomb.metrics.core.monitor;
+
+import com.netflix.servo.monitor.MaxGauge;
+import com.netflix.servo.monitor.MinGauge;
+import com.netflix.servo.monitor.MonitorConfig;
+import com.netflix.servo.monitor.StepCounter;
+
+import io.servicecomb.metrics.core.metric.TimeMetric;
+
+public class TimeMonitor {
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/monitor/TimeMonitor.java]: it doesn't seem to have any time info

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/publish/DefaultDataSource.java [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+
+    List<Long> intervals = Pollers.getPollingIntervals();
+    List<Long> appliedIntervals = new ArrayList<>();
+    for (int index = 0; index < intervals.size(); index++) {
+      int finalIndex = index;
+      long finalInterval = intervals.get(finalIndex) < minPollingTime ? minPollingTime : intervals.get(finalIndex);
+      final Runnable executor = () -> reloadRegistryMetric(finalIndex);
+      Executors.newScheduledThreadPool(1)
+          //for step counter correct work we need poll in time ,otherwise current step will return Datapoint.UNKNOWN (missing last sample)
+          .scheduleWithFixedDelay(executor, 0, (long) ((double) finalInterval / (double) 2), MILLISECONDS);
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/publish/DefaultDataSource.java]: why not `finalInterval * 0.5` to avoid casting to double? and the name executor has a specific meaning in java. preferably to avoid naming runnables as executor

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/registry/MetricsRegistry.java [hunk 27]
```diff
@@ -0,0 +1,23 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.servicecomb.metrics.core.registry;
+
+import io.servicecomb.metrics.core.monitor.RegistryMonitor;
+
+public interface MetricsRegistry {
+  RegistryMonitor getRegistryMonitor();
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/registry/MetricsRegistry.java]: this interface doesn't seem to be necessary

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/registry/DefaultMetricsRegistry.java [hunk 28]
```diff
@@ -0,0 +1,37 @@
...
+
+  private final RegistryMonitor registryMonitor;
+
+  public DefaultMetricsRegistry() {
+    this.registryMonitor = new RegistryMonitor();
+  }
+
+  @Override
+  public RegistryMonitor getRegistryMonitor() {
+    return registryMonitor;
```
Comment (seanyinx) [inline metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/registry/DefaultMetricsRegistry.java]: why not just expose registry monitor as a bean?

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/registry/MetricsRegistry.java (hunk 27 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/registry/MetricsRegistry.java]: 删除，MetricsRegistry在还没有做用户自定义Metrics之前没有需要承载的功能

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/registry/DefaultMetricsRegistry.java (hunk 28 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/registry/DefaultMetricsRegistry.java]: Good Idea

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/publish/DefaultDataSource.java (hunk 26 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/publish/DefaultDataSource.java]: Done

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/monitor/TimeMonitor.java (hunk 25 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/monitor/TimeMonitor.java]: 这个Monitor用于一组计时，改为TimerMonitor

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/monitor/RegistryMonitor.java (hunk 24 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/monitor/RegistryMonitor.java]: Done

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/TimeMetric.java (hunk 22 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/TimeMetric.java]: 当没有请求的时候，时间计数器初试为0，所以0代表无采集数据，merge的时候需要Skip

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/TimeMetric.java (hunk 23 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/TimeMetric.java]: 当没有请求的时候，时间计数器初试为0，所以0代表无采集数据，merge的时候需要Skip，Max不受影响

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/TimeMetric.java (hunk 21 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/metric/TimeMetric.java]: 改为TimerMetric

File: metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationStartedEventListener.java (hunk 20 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/io/servicecomb/metrics/core/event/InvocationStartedEventListener.java]: MetricsRegistry在还没有做用户自定义Metrics之前没有需要承载的功能，直接删除，改为直接注入RegistryMonitor
